File: backend/app/api/v1/routes/system/roles.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends

from app.api.deps import RoleServiceDep, require_permissions
from app.domain.permission import Permission
from app.models.user import User
from app.schemas.role import RoleCreate, RoleResponse, RoleUpdate, StatusResponse

router = APIRouter(
    prefix="/system/roles",
    tags=["System"],
)


@router.get("/")
def list_roles(
    _: Annotated[User, Depends(require_permissions(Permission.ROLES_READ))],
    role_service: RoleServiceDep,
) -> list[RoleResponse]:
    roles = role_service.list_roles()
    return [RoleResponse.model_validate(role) for role in roles]


@router.get("/{role_id}")
def get_role(
    role_id: UUID,
    _: Annotated[User, Depends(require_permissions(Permission.ROLES_READ))],
    role_service: RoleServiceDep,
) -> RoleResponse:
    role = role_service.get_by_id(role_id)
    return RoleResponse.model_validate(role)


@router.post("/", status_code=201)
def create_role(
    role_in: RoleCreate,
    _: Annotated[User, Depends(require_permissions(Permission.ROLES_CREATE))],
    role_service: RoleServiceDep,
) -> RoleResponse:
    role = role_service.create(
        role_name=role_in.role_name,
        permissions=role_in.permissions,
    )
    return RoleResponse.model_validate(role)


@router.patch("/{role_id}")
def update_role(
    role_id: UUID,
    role_in: RoleUpdate,
    _: Annotated[User, Depends(require_permissions(Permission.ROLES_EDIT))],
    role_service: RoleServiceDep,
) -> RoleResponse:
    role = role_service.update(
        role_id=role_id,
        role_name=role_in.role_name,
        permissions=role_in.permissions,
    )
    return RoleResponse.model_validate(role)


@router.delete("/{role_id}")
def delete_role(
    role_id: UUID,
    _: Annotated[User, Depends(require_permissions(Permission.ROLES_DELETE))],
    role_service: RoleServiceDep,
) -> StatusResponse:
    role_service.delete(role_id)
    return StatusResponse()

File: frontend/src/lib/api/system/roles.ts
import { apiFetch } from "@/lib/api/client";
import type { RoleRecord } from "@/lib/api/types";

function mapRole(raw: {
  id: string;
  roleName: string;
  permissions: string[];
  createdAt: string;
  updatedAt: string;
}): RoleRecord {
  return {
    id: raw.id,
    roleName: raw.roleName,
    permissions: raw.permissions,
    createdAt: raw.createdAt,
    updatedAt: raw.updatedAt,
  };
}

export async function listRoles(): Promise<RoleRecord[]> {
  const raw = await apiFetch<
    {
      id: string;
      roleName: string;
      permissions: string[];
      createdAt: string;
      updatedAt: string;
    }[]
  >("/system/roles/");
  return raw.map(mapRole);
}

export async function createRole(input: {
  roleName: string;
  permissions?: string[];
}): Promise<RoleRecord> {
  const raw = await apiFetch<{
    id: string;
    roleName: string;
    permissions: string[];
    createdAt: string;
    updatedAt: string;
  }>("/system/roles/", {
    method: "POST",
    body: {
      roleName: input.roleName,
      permissions: input.permissions ?? [],
    },
  });
  return mapRole(raw);
}

export async function updateRole(
  roleId: string,
  input: { roleName?: string; permissions?: string[] },
): Promise<RoleRecord> {
  const raw = await apiFetch<{
    id: string;
    roleName: string;
    permissions: string[];
    createdAt: string;
    updatedAt: string;
  }>(`/system/roles/${roleId}`, {
    method: "PATCH",
    body: input,
  });
  return mapRole(raw);
}

export async function deleteRole(roleId: string): Promise<void> {
  await apiFetch(`/system/roles/${roleId}`, { method: "DELETE" });
}
